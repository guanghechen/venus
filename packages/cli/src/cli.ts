import { createProgram, mountSubCommandGenerate, mountSubCommandInit } from '.'

const program = createProgram()

// mount sub-command: generate
mountSubCommandGenerate(program)

// mount sub-command: init
mountSubCommandInit(program)

program.parse(process.argv)
