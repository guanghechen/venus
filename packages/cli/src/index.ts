import './env/constant'

export * from './command/generate'
export * from './core/command'
export * from './core/option'
export * from './core/generate/command'
export * from './core/generate/context'
export * from './core/generate/processor'
export * from './env/constant'
export * from './env/logger'
