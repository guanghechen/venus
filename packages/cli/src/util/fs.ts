import fs from 'fs-extra'

/**
 * Check whether if the given filepath exists.
 * @param filepath
 */
export function isExists(filepath: string | null): filepath is string {
  return filepath !== null && fs.existsSync(filepath)
}

/**
 * Check whether if the given filepath exists and that it is a file path.
 * @param filepath
 */
export function isFile(filepath: string | null): filepath is string {
  return isExists(filepath) && fs.statSync(filepath).isFile()
}

/**
 * Check whether if the given filepath exists and that it is a directory path.
 * @param filepath
 */
export function isDirectory(filepath: string | null): boolean {
  return isExists(filepath) && fs.statSync(filepath).isDirectory()
}

/**
 * Ensure the given filepath exists, otherwise, an exception will be thrown.
 * @param filepath
 * @param message
 */
export function ensureExists(
  filepath: string | null,
  message?: string,
): void | never {
  if (isExists(filepath)) return
  throw new ReferenceError(message ?? `Cannot find '${filepath}'.`)
}

/**
 * Ensure the given filepath exists and it is a file path.
 * Otherwise, an exception will be thrown.
 * @param filepath
 * @param message
 */
export function ensureFileExists(
  filepath: string | null,
  message?: string,
): void | never {
  ensureExists(filepath, message)
  if (fs.statSync(filepath!).isFile()) return
  throw new ReferenceError(message ?? `${filepath} is not a file path.`)
}

/**
 * Ensure the given filepath exists and it is a directory path.
 * Otherwise, an exception will be thrown.
 * @param filepath
 * @param message
 */
export function ensureDirectoryExists(
  filepath: string | null,
  message?: string,
): void | never {
  ensureExists(filepath, message)
  if (fs.statSync(filepath!).isDirectory()) return
  throw new ReferenceError(message ?? `${filepath} is not a directory path.`)
}
