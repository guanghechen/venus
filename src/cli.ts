import { createProgram, mountSubCommandGenerate } from '.'

const program = createProgram()

// mount sub-command: generate
mountSubCommandGenerate(program)

program.parse(process.argv)
