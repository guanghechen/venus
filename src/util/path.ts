import { isNonBlankString } from '@guanghechen/option-helper'
import path from 'path'

/**
 * Check if filepath1 under the filepath2.
 *
 * @param filepath1
 * @param filepath2
 * @see https://stackoverflow.com/a/45242825/15760674
 */
export function isUnderThePath(filepath1: string, filepath2: string): boolean {
  const p = path.relative(filepath2, filepath1)
  return isNonBlankString(p) && !p.startsWith('..') && !path.isAbsolute(p)
}

/**
 * Calc the path of filepath relative to cwd.
 *
 * @param filepath          target filepath
 * @param cwd               current workspace directory
 * @param rootDirectory     base directory
 */
export function relativePath(
  filepath: string,
  cwd: string,
  rootDirectory?: string,
): string {
  const absoluteFilepath = path.resolve(cwd, filepath)
  if (rootDirectory != null) {
    if (!isUnderThePath(absoluteFilepath, rootDirectory)) {
      return absoluteFilepath
    }
  }
  return path.relative(cwd, filepath)
}
