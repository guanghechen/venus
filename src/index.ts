import loadCommand from '@/command'
import { doneWithClose } from '@/util/cli'
import { registerCommanderOptions } from '@barusu/chalk-logger'
import program from 'commander'
import manifest from '../package.json'

// 改造 Commander 的 action 方法，使其在执行成功后退出程序（为了退出控制台输入）
{
  const fn = program.Command.prototype as any
  const action = fn.action
  fn.action = function (f: (...args: any[]) => Promise<void>) {
    return action.call(this, doneWithClose(f))
  }
}

program.version(manifest.version)

registerCommanderOptions(program)

loadCommand(program)

program.parse(process.argv)
