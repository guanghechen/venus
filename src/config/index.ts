import { isFile } from '@/util/fs'
import fs from 'fs-extra'
import yaml from 'js-yaml'
import path from 'path'
import { DefaultCreateConfig } from './create'
import { DefaultGenerateConfig } from './generate'
import type { RawDefaultCreateConfig } from './create'
import type { RawDefaultGenerateConfig } from './generate'

const absoluteLocaleConfigPath = path.join(__dirname, 'config.yml')
const localConfigContent = fs.readFileSync(absoluteLocaleConfigPath, 'utf-8')
const localRawConfig = yaml.load(localConfigContent) as RawPartialConfig

/**
 * 配置文件的类型
 */
export interface RawPartialConfig {
  generate: RawDefaultGenerateConfig
  create: RawDefaultCreateConfig
}

/**
 * 获取外部的配置文件
 *
 * @param projectDirectory
 * @param configPath
 */
export const getPartialRawConfig = async (
  projectDirectory: string,
  configPath?: string,
): Promise<RawPartialConfig | undefined> => {
  const absoluteConfigPath = path.resolve(
    projectDirectory,
    configPath != null ? configPath : 'venus.config.yml',
  )
  if (isFile(absoluteConfigPath)) {
    const configContent = fs.readFileSync(absoluteConfigPath, 'utf-8')
    return yaml.load(configContent) as unknown as RawPartialConfig
  }
}

/**
 * 获取子命令 'generate' 的默认选项
 */
export const getDefaultGenerateConfig = (
  partialRawConfig?: RawDefaultGenerateConfig,
): DefaultGenerateConfig => {
  const { generate } = localRawConfig
  return new DefaultGenerateConfig(generate, partialRawConfig)
}

/**
 * 获取子命令 'create/new' 的默认选项
 */
export const getDefaultCreateConfig = (
  partialRawConfig?: RawDefaultCreateConfig,
): DefaultCreateConfig => {
  const { create } = localRawConfig
  return new DefaultCreateConfig(create, partialRawConfig)
}
