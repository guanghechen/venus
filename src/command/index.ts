import path from 'path'
import commander from 'commander'
import { projectConfig } from '@/config/immutable'
import { getPartialRawConfig, RawPartialConfig } from '@/config'
import { ensureFileExist, findNearestTarget } from '@/util/fs-util'
import { coverString } from '@/util/option-util'
import loadGenerateCommand from './generate'
import loadCreateCommand from './create'
import loadRegisterCommand from './register'


export interface GlobalConfig {
  readonly executeDirectory: string
  readonly projectRootDirectory: string
  readonly definitionPhase: string
  readonly encoding: string
  readonly cmakeLists: {
    readonly encoding: string
    readonly filepath: string
  }
}


export default (program: commander.Command) => {
  program
    .option('--cmake-lists <cmakeLists-name>', 'index cmakeLists name. (default is CMakeLists.txt)')
    .option('--encoding <encoding>', 'index encoding of all files.')
    .option('--config-path <config-path>', 'index config path, related with the project root directory where the CMakeLists.txt exists.')
    .option('--no-config', 'don\'t use config.')

  // 挂载 generate 子命令
  loadGenerateCommand(program, getRawPartialConfig('generate'), getGlobalConfig)

  // 挂载 create/new 子命令
  loadCreateCommand(program, getRawPartialConfig('create'), getGlobalConfig)

  // 挂载 register 子命令
  loadRegisterCommand(program, getRawPartialConfig('register'), getGlobalConfig)

  /**
   * 获取外部配置文件
   * @param key
   */
  function getRawPartialConfig(key: 'generate' | 'create' | 'register') {
    return async (specifiedProjectLocatedPath?: string): Promise<RawPartialConfig|any|undefined> => {
      const globalConfig: GlobalConfig = await getGlobalConfig(specifiedProjectLocatedPath)
      if (program.config === false) return
      const partialRawConfig: RawPartialConfig | undefined = await getPartialRawConfig(globalConfig.projectRootDirectory, program.configPath)
      if (partialRawConfig == null) return
      return partialRawConfig[key]
    }
  }

  /**
   * 生成全局选项
   *
   * @param specifiedProjectLocatedPath   该值为绝对路径时才生效，从指定的地方搜索 CMakeLists.txt
   */
  async function getGlobalConfig(specifiedProjectLocatedPath?: string): Promise<GlobalConfig> {
    const { definitionPhase, cmakeLists, encoding } = projectConfig
    const cmakeListsFileName = coverString(cmakeLists.filename, program.cmakeLists)
    const cmakeListsEncoding = coverString(cmakeLists.encoding, program.encoding)

    const executeDirectory = path.resolve()
    if (specifiedProjectLocatedPath != null) specifiedProjectLocatedPath = path.resolve(executeDirectory, specifiedProjectLocatedPath)
    const absoluteCmakeListsPath = specifiedProjectLocatedPath != null
      ? await findNearestTarget(specifiedProjectLocatedPath, cmakeListsFileName)
      : await findNearestTarget(executeDirectory, cmakeListsFileName)

    // 确保 CMakeLists.txt 存在
    await ensureFileExist(absoluteCmakeListsPath, `'${cmakeListsFileName}' is not found.`)

    const projectRootDirectory: string = path.dirname(absoluteCmakeListsPath!)
    return {
      executeDirectory,
      projectRootDirectory,
      definitionPhase,
      encoding,
      cmakeLists: {
        filepath: absoluteCmakeListsPath!,
        encoding: cmakeListsEncoding,
      },
    }
  }
}
