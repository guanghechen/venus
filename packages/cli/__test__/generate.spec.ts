import { fileSnapshot } from '@guanghechen/jest-helper'
import fs from 'fs-extra'
import path from 'path'
import { COMMAND_NAME, createProgram, execSubCommandGenerate } from '../src'

const removeFiles = (dirpath: string, filenames: string[]): void => {
  for (const filename of filenames) {
    const filepath = path.join(dirpath, filename)
    if (fs.existsSync(filepath)) {
      fs.unlinkSync(filepath)
    }
  }
}

describe('simple', function () {
  const caseRoot = path.join(__dirname, 'fixtures/simple')
  const sourceFilepath = path.resolve(caseRoot, 'oj/fake/01.cpp')
  const outputFilenames: string[] = ['venus.cpp']

  afterEach(() => {
    removeFiles(caseRoot, outputFilenames)
  })

  test('basic', async function () {
    const program = createProgram()
    const args = ['', COMMAND_NAME, 'generate', sourceFilepath]
    await execSubCommandGenerate(program, args)
    fileSnapshot(caseRoot, outputFilenames)
  })

  test('uglify', async function () {
    const program = createProgram()
    const args = ['', COMMAND_NAME, 'generate', sourceFilepath, '--uglify']
    await execSubCommandGenerate(program, args)
    fileSnapshot(caseRoot, outputFilenames)
  })
})
