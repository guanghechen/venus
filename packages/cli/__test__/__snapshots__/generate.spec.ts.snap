// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple basic: venus.cpp 1`] = `
"#include <cstdio>
#include <fstream>
#include <functional>
#include <iostream>
using namespace std;


int matchFirst(int lft, int rht, std::function<bool(int)> verify) {
  int rightBound = rht;
  while (lft < rht) {
    int mid = (lft + rht) >> 1;
    if (verify(mid))
      rht = mid;
    else
      lft = mid + 1;
  }
  return rht >= rightBound ? -1 : rht;
}

int main() {
#ifdef GHC_LOCAL
#define in cin
#define out cout
#else
#define in fin
#define out fout

  ofstream fout(\\"test.out\\");
  ifstream fin(\\"test.in\\");
#endif

  int data[5] = { 1, 2, 3, 3, 4 };

  int result = matchFirst(
    0,
    4,
    [&data](int x) -> bool { return data[x] >= 3; }
  );

  
  printf(\\"result: %d\\\\n\\", result);

  char s[] = \\"common /** this is comments */ string\\";
  char t[] = \\"common // this is comments\\";
  return 0;
}
"
`;

exports[`simple pretty: venus.cpp 1`] = `
"#include <cstdio>
#include <fstream>
#include <functional>
#include <iostream>
using namespace std;


/**
 * Find the first matched position.
 */
int matchFirst(int lft, int rht, std::function<bool(int)> verify) {
  int rightBound = rht;
  while (lft < rht) {
    int mid = (lft + rht) >> 1;
    if (verify(mid))
      rht = mid;
    else
      lft = mid + 1;
  }
  return rht >= rightBound ? -1 : rht;
}

int main() {
#ifdef GHC_LOCAL
#define in cin
#define out cout
#else
#define in fin
#define out fout

  ofstream fout(\\"test.out\\");
  ifstream fin(\\"test.in\\");
#endif

  int data[5] = { 1, 2, 3, 3, 4 };

  /**
   * Multiple
   * line
   * Comments
   */
  int result = matchFirst(
    0,
    4,
    [&data](int x) -> bool { return data[x] >= 3; }
  );

  // Single line comments
  printf(\\"result: %d\\\\n\\", result);

  char s[] = \\"common /** this is comments */ string\\";
  char t[] = \\"common // this is comments\\";
  return 0;
}
"
`;

exports[`simple uglify: venus.cpp 1`] = `
"#include<cstdio>
#include <fstream>
#include <functional>
#include <iostream>
using namespace std;
int matchFirst(int lft,int rht,std::function<bool(int)>verify){int rightBound=rht;while(lft<rht){int mid=(lft+rht)>>1;if(verify(mid))rht=mid;else lft=mid+1;}return rht>=rightBound?-1:rht;}int main(){#ifdef GHC_LOCAL
#define in cin
#define out cout
#else
#define in fin
#define out fout
ofstream fout(\\"test.out\\");ifstream fin(\\"test.in\\");#endif
int data[5]={1,2,3,3,4};int result=matchFirst(0,4,[&data](int x)->bool{return data[x]>=3;});printf(\\"result: %d\\\\n\\",result);char s[]=\\"common /** this is comments */ string\\";char t[]=\\"common // this is comments\\";return 0;}"
`;
